### PR TITLE
Use bool for TT stages in movepick

### PR DESCRIPTION
**Pull request by protonspring**
This is a non-functional simplification.  If we use a bool as a flag for using the ttMove or not, we can remove all of the TT stages.  The bool (useTTM) is set once and checked once or twice (if there was a ttMove) in the INIT stages.

This decreases the jump table size, decreases the executable size, and is a bit faster on my machines (< 1%).  This patch is identical to the tested code, but with a better variable name and added comments for clarity.

STC
LLR: 2.95 (-2.94,2.94) {-1.50,0.50}
Total: 80802 W: 15670 L: 15602 D: 49530
Ptnml(0-2): 1337, 9181, 19268, 9307, 1308
https://tests.stockfishchess.org/tests/view/5e7e7705e42a5c3b3ca2ec71

### CODE REVIEW

File: src/movepick.cpp [hunk 1]
```diff
@@ -63,9 +63,8 @@ MovePicker::MovePicker(const Position& p, Move ttm, Depth d, const ButterflyHist
 
   assert(d > 0);
 
-  stage = pos.checkers() ? EVASION_TT : MAIN_TT;
-  ttMove = ttm && pos.pseudo_legal(ttm) ? ttm : MOVE_NONE;
-  stage += (ttMove == MOVE_NONE);
+  stage = pos.checkers() ? EVASION_INIT : CAPTURE_INIT;
+  useTTM = ((ttMove = ttm && pos.pseudo_legal(ttm) ? ttm : MOVE_NONE));
```
Comment (vondele) [inline src/movepick.cpp]: are all parenthesis needed here?

File: src/movepick.cpp [hunk 2]
```diff
@@ -75,11 +74,10 @@ MovePicker::MovePicker(const Position& p, Move ttm, Depth d, const ButterflyHist
...
   assert(d <= 0);
 
-  stage = pos.checkers() ? EVASION_TT : QSEARCH_TT;
-  ttMove =   ttm
+  stage = pos.checkers() ? EVASION_INIT : QCAPTURE_INIT;
+  useTTM = ((ttMove =   (ttm
           && (depth > DEPTH_QS_RECAPTURES || to_sq(ttm) == recaptureSquare)
-          && pos.pseudo_legal(ttm) ? ttm : MOVE_NONE;
-  stage += (ttMove == MOVE_NONE);
+          && pos.pseudo_legal(ttm) ? ttm : MOVE_NONE)));
```
Comment (vondele) [inline src/movepick.cpp]: idem

File: src/movepick.cpp [hunk 3]
```diff
@@ -89,12 +87,11 @@ MovePicker::MovePicker(const Position& p, Move ttm, Value th, const CapturePiece
...
-          && pos.capture(ttm)
-          && pos.pseudo_legal(ttm)
-          && pos.see_ge(ttm, threshold) ? ttm : MOVE_NONE;
-  stage += (ttMove == MOVE_NONE);
+  stage = PROBCUT_INIT;
+  useTTM = ((ttMove =   ttm
+                      && pos.capture(ttm)
+                      && pos.pseudo_legal(ttm)
+                      && pos.see_ge(ttm, threshold) ? ttm : MOVE_NONE));
 }
```
Comment (vondele) [inline src/movepick.cpp]: idem

File: src/movepick.cpp [hunk 4]
```diff
@@ -156,16 +153,11 @@ Move MovePicker::next_move(bool skipQuiets) {
...
-  case EVASION_TT:
-  case QSEARCH_TT:
-  case PROBCUT_TT:
-      ++stage;
-      return ttMove;
-
   case CAPTURE_INIT:
   case PROBCUT_INIT:
   case QCAPTURE_INIT:
+      if (!(useTTM = !useTTM)) return ttMove; //if set, unset and return ttm
```
Comment (vondele) [inline src/movepick.cpp]: new line for return.
can't we rewrite that '(!(useTTM = !useTTM)) ' more cleanly, it is hard to parse.

Comment (protonspring) [reply src/movepick.cpp]: It can be more clear, but will probably require brackets like:
if (useTTM)
{
     useTTM = !useTTM;
     return ttMove;
}

Is that better?

File: src/movepick.cpp (hunk 1 above)
Comment (protonspring) [reply src/movepick.cpp]: If I remove the extra set, the compiler complains about using an assignment in a truth check or something like that.

File: src/movepick.cpp (hunk 4 above)
Comment (vondele) [reply src/movepick.cpp]: easier to read... 

makes one wonder if this is really a simplification.

Comment (protonspring) [reply src/movepick.cpp]: I could also do a simple method that does the same.  it would make the jump table quite easy to read, but a few more lines.  To me, this is still better (more simple) that a bunch of TT stages.

File: src/movepick.cpp (hunk 1 above)
Comment (protonspring) [reply src/movepick.cpp]: I was able to remove them.

File: src/movepick.cpp (hunk 4 above)
Comment (protonspring) [reply src/movepick.cpp]: I added a new line after the check and added a comment.  To me, this seems pretty easy to read/understand.

File: src/movepick.cpp [hunk 5]
```diff
@@ -63,9 +62,8 @@ MovePicker::MovePicker(const Position& p, Move ttm, Depth d, const ButterflyHist
 
   assert(d > 0);
 
-  stage = pos.checkers() ? EVASION_TT : MAIN_TT;
-  ttMove = ttm && pos.pseudo_legal(ttm) ? ttm : MOVE_NONE;
-  stage += (ttMove == MOVE_NONE);
+  stage = pos.checkers() ? EVASION_INIT : CAPTURE_INIT;
+  useTTM = (ttMove = ttm && pos.pseudo_legal(ttm) ? ttm : MOVE_NONE);
```
Comment (vondele) [inline src/movepick.cpp]: Do we still need the ternary operator here?

File: src/movepick.cpp [hunk 6]
```diff
@@ -75,11 +73,10 @@ MovePicker::MovePicker(const Position& p, Move ttm, Depth d, const ButterflyHist
...
   assert(d <= 0);
 
-  stage = pos.checkers() ? EVASION_TT : QSEARCH_TT;
-  ttMove =   ttm
+  stage = pos.checkers() ? EVASION_INIT : QCAPTURE_INIT;
+  useTTM = (ttMove =   (ttm
           && (depth > DEPTH_QS_RECAPTURES || to_sq(ttm) == recaptureSquare)
-          && pos.pseudo_legal(ttm) ? ttm : MOVE_NONE;
-  stage += (ttMove == MOVE_NONE);
+          && pos.pseudo_legal(ttm) ? ttm : MOVE_NONE));
```
Comment (vondele) [inline src/movepick.cpp]: idem

File: src/movepick.cpp [hunk 7]
```diff
@@ -89,12 +86,10 @@ MovePicker::MovePicker(const Position& p, Move ttm, Value th, const CapturePiece
...
-  stage = PROBCUT_TT;
-  ttMove =   ttm
-          && pos.capture(ttm)
-          && pos.pseudo_legal(ttm)
-          && pos.see_ge(ttm, threshold) ? ttm : MOVE_NONE;
-  stage += (ttMove == MOVE_NONE);
+  stage = PROBCUT_INIT;
+  useTTM = (ttMove = ttm && pos.capture(ttm)
+                         && pos.pseudo_legal(ttm)
+                         && pos.see_ge(ttm, threshold) ? ttm : MOVE_NONE);
```
Comment (vondele) [inline src/movepick.cpp]: idem

File: src/movepick.cpp [hunk 8]
```diff
@@ -75,11 +73,10 @@ MovePicker::MovePicker(const Position& p, Move ttm, Depth d, const ButterflyHist
...
 
-  stage = pos.checkers() ? EVASION_TT : QSEARCH_TT;
-  ttMove =   ttm
-          && (depth > DEPTH_QS_RECAPTURES || to_sq(ttm) == recaptureSquare)
-          && pos.pseudo_legal(ttm) ? ttm : MOVE_NONE;
-  stage += (ttMove == MOVE_NONE);
+  stage = pos.checkers() ? EVASION_INIT : QCAPTURE_INIT;
+  useTTM = (ttMove = (ttm && (depth > DEPTH_QS_RECAPTURES
+                              || to_sq(ttm) == recaptureSquare)
+                          && pos.pseudo_legal(ttm) ? ttm : MOVE_NONE));
```
Comment (vondele) [inline src/movepick.cpp]: idem

File: src/movepick.cpp (hunk 5 above)
Comment (protonspring) [reply src/movepick.cpp]: Excellent observation.  Yes.  I think that can go away now as well.  Just a sec.

Comment (protonspring) [reply src/movepick.cpp]: Darn. . doesn't look like it.  If the given ttm is not legal, ttMove has to be set to MOVE_NONE or the bench changes.

Comment (vondele) [reply src/movepick.cpp]: also, since we now unconditionally assign to ttMove, it can be done in the initializer list in all three constructors, which further simplifies the conditional.

Comment (vondele) [reply src/movepick.cpp]: OK, so my last comment is not quite right.

Comment (vondele) [reply src/movepick.cpp]: actually, I'm not sure why. ttm shouldn't be used if it is not pseudo-legal? It might result in an other generated move (that is not pseudo legal either, being skipped later in the movepicker)?

Comment (protonspring) [reply src/movepick.cpp]: You may be right.  Should I further simplify this way and test it?

Comment (protonspring) [reply src/movepick.cpp]: I also considered ttMove = ttm * (conditional checks).  If false, ttMove would be MOVE_NONE.  If true, ttm.  Perhaps that is too cryptic.

Comment (vondele) [reply src/movepick.cpp]: That latter suggestion is too cryptic. Before testing on fishtest, you can maybe try to understand if my suggested reason is indeed correct. Afterwards, as this becomes a functional change, it should be retested.

Comment (protonspring) [reply src/movepick.cpp]: OK.  Will do.

Comment (protonspring) [reply src/movepick.cpp]: In master, if ttm fails to meet the conditions (e.g., pseudo_legal), ttMove is set to MOVE_NONE.  In this case, we may still return the ttMove (even if not pseudo_legal) because we no longer track what it was.

Thus, it seems BETTER to always assign ttm to ttMove.  This way, we NEVER return the ttMove regardless of whether it met the prescribed conditions (i.e., pseudo_legal, etc.).

If that seems right to you, should I test that on the framework?

Comment (vondele) [reply src/movepick.cpp]: I think that's correct indeed. Note that in the probcut constructor stage could also be set in the initializer list.

Comment (protonspring) [reply src/movepick.cpp]: https://tests.stockfishchess.org/tests/view/5e821a76e42a5c3b3ca2edfa

Comment (protonspring) [reply src/movepick.cpp]: Test failed, so I'm not sure we can remove this ternary operator. . yet.  I'll keep thinking about it and make a different PR if I think of something.

**Comment by vondele** [comment]
upon closer look, I don't think this is a simplification. The introduce bool is a little more state, that is similar to the stages being removed. Line count is also the same if we remove use separate lines for the assignments. I also don't measure any speedups locally. Ok to close?

**Comment by protonspring** [comment]
Sure. NP.
